core/drivers/clk/clk.c

```diff
@@ -0,0 +1,273 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2021, Bootlin
+ */
+
+#include <drivers/clk.h>
+#include <kernel/boot.h>
+#include <kernel/panic.h>
+#include <kernel/spinlock.h>
+#include <libfdt.h>
+#include <malloc.h>
+#include <stddef.h>
+
+/* Global clock tree lock */
+static unsigned int clk_lock = SPINLOCK_UNLOCK;
+
+struct clk *clk_alloc(const char *name, const struct clk_ops *ops,
+		      struct clk **parent_clks, size_t parent_count)
+{
+	struct clk *clk = NULL;
+	size_t parent = 0;
+
+	clk = calloc(1, sizeof(*clk) + parent_count * sizeof(clk));
+	if (!clk)
+		return NULL;
+
+	clk->num_parents = parent_count;
+	for (parent = 0; parent < parent_count; parent++)
+		clk->parents[parent] = parent_clks[parent];
+
+	clk->name = name;
+	clk->ops = ops;
+	refcount_set(&clk->enabled_count, 0);
+
+	return clk;
+}
+
+void clk_free(struct clk *clk)
+{
+	free(clk);
+}
+
+static bool clk_check(struct clk *clk)
```

Comment:
add `__maybe_unused`

Reply:
Ok, didn't think of that when assert are disabled.